### DISCUSSION TITLE
Applying changes in a loop (one by one) - is it possible ?

### DISCUSSION THREAD

**Discussion by przemolb** [aws/aws-cdk#26545] [Q&A]
Hello,

I have the following use case and given `cdk synth` and `cdk deploy` I am not sure if it is possible with AWS CDK.
I want to create X  (where X ~ 100) number of the same objects (i.e., ECS clusters). While bootstrapping of one object is not a problem, I wonder how can I create (or modify) all those objects in a loop, one by one.

This is one scenario:
1. create an ECS cluster "ECS-1" with a task definition
2. run some basic test (does an agent running as a container register correctly and proceed to the next step only if it succeeded)
3. create ECS cluster "ECS-2"
4. run some basic test (does an agent running as a container register correctly and proceed to the next step only if it succeeded)
5. ... carry on in a loop ...

This is another scenario:
1. modify Task definition in ECS-1 cluster (upgrade image version from X to Y)
2. run some tests to verify the image is working correctly
3. modify Task definition in ECS-2 cluster (upgrade image version from X to Y)
4. run some tests to verify the image is working correctly
6. ... carry on in a loop for the other ECS clusters ...

Given that after each change, we have to run `cdk synth` and `cdk deploy` is the above scenario possible while using AWS CDK ?

(if that matters - I was going to use Golang for that)

**Comment by przemolb** [comment]
Any hints on whether it is possible with AWS CDK ?

**Reply by oshea00** [reply]
This would not be practical using CDK. It's still IaC under the covers and the way it's done is to express the desired state of the ending architecture configuration and then creating that state "all at once". 

What you describe would be better done (in AWS) using an Autoscaling Group definition that's allowed to grow from a single EC2 instance to some maximum number based on scaling policies. You could describe that configuration using CDK and then deploy it.  The dynamic behavior of the cluster is determined by the policies you configure.

You might use your CDK logic to adjust the polices based on any rules you'd like to code up - but the actual deployment is "all at once".

**Reply by przemolb** [reply]
Well, in case of EC2 instances, it (ASG) might work. But I am talking about ECS.
Each ECS cluster will be in a different, dedicated AWS account.

It looks like AWS CDK is not very well suited to meet this type of requirement?

**Reply by oshea00** [reply]
CDK helps to generate terraform and this can make the expression of an architecture perhaps easier for developers to deal with - maybe even modularizing components, but at the end of the day it's still terraform.

**Reply by przemolb** [reply]
@oshea00 you are talking about `cdktf`. This thread is about `AWS CDK`.

**Reply by oshea00** [reply]
Oh ha ha. Sorry, yeah, I use both frameworks and I have had Terraform on the brain lately. Cloudformation CDK, Terraform CDKTF, either way, same answer as they are based on the same technology - IAC.

If you did something like deploy a stack, do some testing, generate some more IaC, deploy it again, do some more testing, etc. You eventually work your way up to a final configuration that you could roll back at any point, but each deploy would just be adding to the final state. Not sure what the advantage would be.

**Reply by przemolb** [reply]
The advantage of doing the changes one by one is that I can stop and investigate once something wrong has happened. I.e. modify  one ECS cluster, if it succeeded then carry on with the next one. If not then stop and let me investigate.
